enterprise/frontend/src/embedding-sdk/bin/get-public-components.ts

```diff
@@ -0,0 +1,92 @@
+import {
+  type ExportSpecifier,
+  type Identifier,
+  type Node,
+  Project,
+  SyntaxKind,
+} from "ts-morph";
+
+import { isNotNull } from "metabase/lib/types";
+
+import type { ComponentDefinition } from "./generate-nextjs-compat";
+
+export function getPublicComponents() {
+  const project = new Project();
+  const sdkPublicIndexSourceFile = project.addSourceFileAtPath(
+    "enterprise/frontend/src/embedding-sdk/components/public/index.ts",
+  );
+  project.addSourceFilesAtPaths(
+    "enterprise/frontend/src/embedding-sdk/components/public/**/*{.ts,.tsx}",
+  );
+
+  const componentForNextJsCompatList: ComponentDefinition[] = [];
+
+  sdkPublicIndexSourceFile.forEachDescendant(node => {
+    switch (node.getKind()) {
+      case SyntaxKind.ExportSpecifier: {
+        if (!isType(node as ExportSpecifier) && isReactComponent(node)) {
+          const componentForNextJsCompat: ComponentDefinition = {
+            mainComponent: node.getText(),
+            subComponents: [],
+          };
+          componentForNextJsCompatList.push(componentForNextJsCompat);
+
+          componentForNextJsCompat.subComponents = findSubComponents(
+            node as ExportSpecifier,
+          );
+        }
+        break;
+      }
+    }
+  });
+
+  return componentForNextJsCompatList;
+}
+
+function findSubComponents(node: ExportSpecifier) {
+  const sdkComponentReferences = (
+    node.getFirstChild() as Identifier
+  ).findReferencesAsNodes();
+
+  return sdkComponentReferences
+    .map(reference => {
+      if (
+        // Find all `InteractiveQuestion.Xxx` references.
+        reference.getParent()?.getKind() ===
+          SyntaxKind.PropertyAccessExpression &&
+        // Don't include `<InteractiveQuestion.Filter />` in a reference, since they're duplicates.
+        !isJsxElement(reference.getParent()?.getParent())
+      ) {
+        /**
+         * [     reference    ]            -> Identifier
+         * [     reference.parent       ]  -> PropertyAccessExpression
+         *                     [lastChild] -> Identifier
+         * InteractiveQuestion.BackButton = BackButton;
+         */
+        const subComponent = reference.getParent()?.getLastChild()?.getText();
+        return subComponent;
+      }
+    })
+    .filter(isNotNull);
+}
+
+function isReactComponent(node: Node) {
+  return node.getText().match(/^[A-Z]/);
+}
+
+/**
+ *
+ * @returns {boolean} Whether the export node is a type
+ * e.g. `export type { Foo } from './Foo'` or `export { type Foo } from './Foo'`
+ */
+function isType(node: ExportSpecifier) {
+  return node.isTypeOnly() || node.getParent()?.getParent().isTypeOnly();
+}
+
+function isJsxElement(node?: Node) {
+  if (!node) {
+    return false;
+  }
+  const kind = node.getKind();
+  return kind >= SyntaxKind.JsxElement && kind <= SyntaxKind.JsxClosingFragment;
```

Comment:
Since these types are related and are next to each other, we can just compare it easier as a number.

https://github.com/dsherret/ts-morph/blob/6c092c8a760b38e9ed3bebdc75f6d6ad654762ea/packages/common/lib/typescript.d.ts#L3950-L3956

Reply:
This seems to be a file that they extended from typescript. I guess the part I'm referring to hasn't been modified by the author.

https://github.com/microsoft/TypeScript/blob/caf1aee269d1660b4d2a8b555c2d602c97cb28d7/src/compiler/types.ts#L360-L366

I checked that the order matches.

However, I like the explicitness of your idea. I'll do that instead.

Reply:
@npretto fixed in 5c07db0817b957aee72afb058fc517834ee69473